lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -6759,145 +6793,144 @@ class PullbackEmitter final : public SILInstructionVisitor<PullbackEmitter> {
-                    makeConcreteAdjointValue(newAdjValue), si->getLoc());
-    builder.createDeallocStack(ai->getLoc(), subscriptBuffer);
+    addAdjointValue(si->getParent(), src, makeConcreteAdjointValue(eltAdjValue),
+                    si->getLoc());
+    builder.createDeallocStack(si->getLoc(), eltAdjBuffer);
   }
 
-  void accumulateArrayTangentSubscriptIndirect(
-      ApplyInst *ai, CopyAddrInst *cai, AllocStackInst *subscriptBuffer) {
-    addToAdjointBuffer(cai->getParent(), cai->getSrc(), subscriptBuffer,
+  /// Accumulate array element adjoint buffer into `copy_addr` source.
+  void accumulateArrayElementAdjointIndirect(CopyAddrInst *cai,
+                                             AllocStackInst *eltAdjBuffer) {
+    addToAdjointBuffer(cai->getParent(), cai->getSrc(), eltAdjBuffer,
                        cai->getLoc());
-    builder.emitDestroyAddrAndFold(cai->getLoc(), subscriptBuffer);
-    builder.createDeallocStack(ai->getLoc(), subscriptBuffer);
+    builder.emitDestroyAddrAndFold(cai->getLoc(), eltAdjBuffer);
+    builder.createDeallocStack(cai->getLoc(), eltAdjBuffer);
   }
 
-  void visitArrayInitialization(ApplyInst *ai) {
-    LLVM_DEBUG(getADDebugStream() << "Visiting array initialization:\n" << *ai);
-    SILValue adjointArray;
-    SILValue fnRef;
-    CanGenericSignature genericSig;
-    for (auto use : ai->getUses()) {
-      auto *dti = dyn_cast<DestructureTupleInst>(use->getUser());
-      if (!dti) continue;
-      // The first tuple field of the return value is the `Array`.
-      adjointArray = getAdjointValue(ai->getParent(), dti->getResult(0))
-          .getConcreteValue();
-      assert(adjointArray && "Array does not have adjoint value");
-      auto astType = adjointArray->getType().getASTType();
-      auto typeDecl = astType->getStructOrBoundGenericStruct();
-      auto subscriptDecl = cast<SubscriptDecl>(typeDecl->lookupDirect(
-          DeclBaseName::createSubscript()).front());
-      auto subscriptGet = subscriptDecl->getAccessor(AccessorKind::Get);
-      SILDeclRef subscriptRef(subscriptGet, SILDeclRef::Kind::Func);
-      auto fnBuilder = SILOptFunctionBuilder(getContext().getTransform());
-      auto fn = fnBuilder.getOrCreateFunction(
-          ai->getLoc(), subscriptRef, NotForDefinition);
-      genericSig = fn->getLoweredFunctionType()->getSubstGenericSignature();
-      fnRef = builder.createFunctionRef(ai->getLoc(), fn);
-    }
-    assert(adjointArray && "Array does not have adjoint value");
-    assert(genericSig && "No generic signature");
-    assert(fnRef && "Could not create `function_ref`");
-    // Two loops because the `tuple_extract` instructions can be reached in
-    // either order.
+  /// Given a `store` or `copy_addr` instruction whose destination is an element
+  /// address from an `array.uninitialized_intrinsic` application, accumulate
+  /// array element adjoint into the source's adjoint.
+  void accumulateArrayElementAdjoint(SILInstruction *inst) {
```

Comment:
(Can be revisited later if you'd rather submit this improvement now instead of spending more time.)

Can this be combined with the logic for the normal stores/copies, by adding an array case to `getAdjointProjection`? If doable, this would slightly reduce the amount of code and would also take advantage of the recursiveness of `getAdjointProjection` to handle the tuple_element_addr/struct_element_addr case that I keep mentioning.

Reply:
> Can this be combined with the logic for the normal stores/copies, by adding an array case to `getAdjointProjection`? If doable, this would slightly reduce the amount of code and would also take advantage of the recursiveness of `getAdjointProjection` to handle the tuple_element_addr/struct_element_addr case that I keep mentioning.

Good suggestion! I filed TF-976 to track this improvement.